### PR TITLE
Speed up DateHeaderValueManager

### PR DESCRIPTION
**Pull request by tenor**
This is the PR for #430 

I removed PumpTimer() and replaced it with StartTimer() and introduced a new StopTimer() method which pauses the timer instead of disposing it.

It's about 2 to 3 times faster than the existing manager and it's more precise because it polls the RTC every 200ms.

Here are some notes I made while working on the change:

1. The older `ISystemClock` used `DateTimeOffset.UtcNow`. That property is about 10 times slower than `DateTime.UtcNow`. I've changed `ISystemClock` to use `DateTime.UtcNow` instead.
Is there a reason why `DateTimeOffset` is preferred? I can change it back without having any effect on the performance. It's only called in the timer loop.
*Update: I've changed it back to DateTimeOffset due to failing unit tests.*

2. The `_activeDateBytes` flag toggles between reading `_dateBytes0` and `_dateBytes1`. I understand this is done to prevent allocations but it looks what's needed here is a buffer pool, so you can say `pool.GetBuffer(size)` and work with the provided buffer.

Here are the results for the speed test. You can download the speed test project from [here](https://www.dropbox.com/s/dxnacgbzwxfgbfd/ImprovedDateHeaderTime.zip?dl=0)

On dnx x64:
```
Testing Old Manager....
7 threads @ 50,000,000 calls per thread.
Total Time: 00:00:32.0746731


Testing New Manager....
7 threads @ 50,000,000 calls per thread.
Total Time: 00:00:11.7850176
Test Complete!
```

On dnx x86:
```
Testing Old Manager....
7 threads @ 50,000,000 calls per thread.
Total Time: 00:00:21.1245817


Testing New Manager....
7 threads @ 50,000,000 calls per thread.
Total Time: 00:00:11.2784120
Test Complete!


```

### CODE REVIEW

File: src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs [hunk 1]
```diff
@@ -34,9 +37,8 @@ public DateHeaderValueManager()
             : this(
                   systemClock: new SystemClock(),
                   timeWithoutRequestsUntilIdle: TimeSpan.FromSeconds(10),
-                  timerInterval: TimeSpan.FromSeconds(1))
+                  timerInterval: TimeSpan.FromMilliseconds(200))
```
Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs]: Not sure running 5 times as often is necessary? As it doesn't need to be highly accurate (reasons given in issue)

**Comment by davidfowl** [comment]
@tenor How does this affect the RPS?

File: src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs (hunk 1 above)
Comment (tenor) [inline src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs]: The timerInterval only affects the precision and has no effect on the performance gains.
GetDatexxx method calls will still be 2-3 times faster regardless of what is set as the timer interval.

**Comment by tenor** [comment]
@davidfowl I expect to see some slight RPS gains but nothing significant.
I only removed one impediment in the request path but there are probably other bottlenecks in the request path.

**Comment by davidfowl** [comment]
If there's no real gain in RPS or overall memory reduction then there's no point.

**Comment by tenor** [comment]
Say you have a pipe with multiple clogs. Someone clears one clog but that fix is rejected because the overall flow is not impacted, then that pipe will never get de-clogged.

The cool thing about open source is that over time, a lot of those bottlenecks tend to be found and fixed.

File: src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs (hunk 1 above)
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs]: I would remove this change from this PR then. If you think it's worth it to increase the precision of the date header, just submit another PR with this one line change.

File: src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs [hunk 2]
```diff
@@ -57,103 +61,159 @@ public DateHeaderValueManager()
...
+            _activeDateBytes = !_activeDateBytes;
+        }
+
+        private void WriteLongThreadSafe(ref long location, long value)
         {
-            if (_dateValueTimer != null)
+            Interlocked.Exchange(ref location, value);
+        }
+
+        private long ReadLongThreadSafe(ref long location)
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs]: Do we know if this is more efficient on 64-bit machines than simply calling Interlocked.Read?

File: src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs [hunk 3]
```diff
@@ -57,103 +61,159 @@ public DateHeaderValueManager()
...
+        /// <param name="value">A DateTimeOffset value</param>
+        private void SetDateValues(DateTimeOffset value)
+        {
+            // See http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.18 for required format of Date header
+            _dateValue = value.ToString(Constants.RFC1123DateFormat);
+            Encoding.ASCII.GetBytes(_dateValue, 0, _dateValue.Length, !_activeDateBytes ? _dateBytes0 : _dateBytes1, "\r\nDate: ".Length);
+            _activeDateBytes = !_activeDateBytes;
+        }
+
+        private void WriteLongThreadSafe(ref long location, long value)
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs]: I think Interlocked.Exchange can be called directly.

File: src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs [hunk 4]
```diff
@@ -57,103 +61,159 @@ public DateHeaderValueManager()
...
-            PumpTimer();
-
-            // See https://msdn.microsoft.com/en-us/library/az4se3k1(v=vs.110).aspx#RFC1123 for info on the format
-            // string used here.
-            // The null-coalesce here is to protect against returning null after Dispose() is called, at which
-            // point _dateValue will be null forever after.
-            return _dateValue ?? _systemClock.UtcNow.ToString(Constants.RFC1123DateFormat);
+            _hadRequestsSinceLastTimerTick = true;
+            PrepareDateValues();
+            return _dateValue;
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs]: So after this is disposed, GetDateHeaderValue() will forever return the value as it was when it was disposed.

This seems fine to me, but given the old comment, I want to verify this is OK. @DamianEdwards?

**Comment by halter73** [comment]
OK. I just saw the RPS comments. @davidfowl Are you planning on closing this?

File: src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs (hunk 1 above)
Comment (tenor) [inline src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs]: @halter73 Sure, I'll change this to one second and submit another PR that sets it to 400ms.

File: src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs (hunk 4 above)
Comment (tenor) [inline src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs]: Yes, it will, however,  `PrepareDateValues()` will return the date from the real time clock if `_isDisposed` is true. The older code behaved this way too.

File: src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs (hunk 2 above)
Comment (tenor) [inline src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs]: Yes, Interlocked.Read is mostly useful for 32-bit systems. It eliminates the possibility of having a torn read, which can't happen when reading a 64 bit value on a 64 bit machine (since .NET aligns the word properly).

File: src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs (hunk 3 above)
Comment (tenor) [inline src/Microsoft.AspNet.Server.Kestrel/Http/DateHeaderValueManager.cs]: The ReadLong/WriteLongThreadSafe methods are really there to remind anyone modifying this class that they need to use Interlocked methods when getting/setting those fields.
Currently, there'll only be one thread reading/writing those fields at any time (in the TimerLoop), there's no need to use InterlockedXXX to access them since it's just one thread.

I added those methods because you can't be too paranoid with multi-threaded code. In fact I should add `[MethodImpl(MethodImplOptions.NoInlining)]` attribute to both methods, even though I don't think they are candidates for inlining.

**Comment by halter73** [comment]
I talked to @davidfowl in person and he doesn't really mind if I merge this. I want include this, since this a pretty limited change, and you produced some nice micro benchmark numbers.

I do have some style changes that I would like to make to this. Normally I would just comment on the relevant lines in the PR, but I made the changes myself so I could ensure that they wouldn't hurt performance (namely removing WriteLongThreadSafe and ReadLongThreadSafe).

Here are the proposed changes: https://github.com/aspnet/KestrelHttpServer/commit/5580181dfe09d71747ea559f04402074f5ef0806

I ran your speed test project and saw no perf hit on both 32-bit and 64-bit dnx.

@tenor Do my proposed changes look good to you? If so, I can just merge this with my changes.

P.S. In the future it would be better to create a github repo instead of linking to a zip archive, but it's no big deal.

**Comment by tenor** [comment]
@halter73 Your changes look good to me. Please merge. :smile:

**Comment by halter73** [comment]
Done! Thanks again.

**Comment by tenor** [comment]
You're welcome! Thanks for looking into the PR.

**Comment by davidfowl** [comment]
@tenor so you're going to send more now right?

**Comment by tenor** [comment]
@davidfowl Sure, if I find more stuff. :smirk:
